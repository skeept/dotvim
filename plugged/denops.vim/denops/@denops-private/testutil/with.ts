import { ADDR_ENV_NAME } from "./cli.ts";
import { getConfig } from "./conf.ts";

const script = new URL("./cli.ts", import.meta.url);

export type Fn<T> = (
  reader: ReadableStream<Uint8Array>,
  writer: WritableStream<Uint8Array>,
) => Promise<T>;

export type WithOptions<T> = {
  fn: Fn<T>;
  prelude?: string[];
  postlude?: string[];
  env?: Record<string, string>;
};

export function withVim<T>(
  options: WithOptions<T>,
): Promise<T> {
  const conf = getConfig();
  const exec = Deno.execPath();
  const commands = [
    ...(options.prelude ?? []),
    `set runtimepath^=${conf.denopsPath}`,
    `let g:denops_test_channel = job_start(['${exec}', 'run', '--allow-all', '${script}'], {'mode': 'json', 'err_mode': 'nl'})`,
    ...(options.postlude ?? []),
  ];
  const cmd = conf.vimExecutable;
  const args = [
    "-u",
    "NONE", // Disable vimrc, plugins, defaults.vim
    "-i",
    "NONE", // Disable viminfo
    "-n", // Disable swap file
    "-N", // Disable compatible mode
    "-X", // Disable xterm
    "-e", // Start Vim in Ex mode
    "-s", // Silent or batch mode
    ...commands.flatMap((c) => ["-c", c]),
  ];
  return withProcess(cmd, args, conf.verbose, options);
}

export function withNeovim<T>(
  options: WithOptions<T>,
): Promise<T> {
  const conf = getConfig();
  const exec = Deno.execPath();
  const commands = [
    ...(options.prelude ?? []),
    `set runtimepath^=${conf.denopsPath}`,
    `let g:denops_test_channel = jobstart(['${exec}', 'run', '--allow-all', '${script}'], {'rpc': v:true})`,
    ...(options.postlude ?? []),
  ];
  const cmd = conf.nvimExecutable;
  const args = [
    "--clean",
    "--embed",
    "--headless",
    "-n",
    ...commands.flatMap((c) => ["-c", c]),
  ];
  return withProcess(cmd, args, conf.verbose, options);
}

async function withProcess<T>(
  cmd: string,
  args: string[],
  verbose: boolean,
  { fn, env }: WithOptions<T>,
): Promise<T> {
  const listener = Deno.listen({
    hostname: "127.0.0.1",
    port: 0, // Automatically select free port
  });
  if (verbose) {
    args.unshift("--cmd", "redir >> /dev/stdout");
  }
  const command = new Deno.Command(cmd, {
    args,
    stdin: "piped",
    stdout: verbose ? "inherit" : "null",
    stderr: verbose ? "inherit" : "null",
    env: {
      ...env,
      [ADDR_ENV_NAME]: JSON.stringify(listener.addr),
    },
  });
  const proc = command.spawn();
  const conn = await listener.accept();
  try {
    return await fn(conn.readable, conn.writable);
  } finally {
    listener.close();
    proc.kill();
    await Promise.all([
      proc.stdin?.close(),
      proc.output(),
    ]);
  }
}
