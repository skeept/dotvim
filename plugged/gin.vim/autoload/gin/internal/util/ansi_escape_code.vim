" Ref: https://github.com/w0ng/vim-hybrid
let s:terminal_ansi_colors = [
      \ '#282a2e',
      \ '#a54242',
      \ '#8c9440',
      \ '#de935f',
      \ '#5f819d',
      \ '#85678f',
      \ '#5e8d87',
      \ '#707880',
      \ '#373b41',
      \ '#cc6666',
      \ '#b5bd68',
      \ '#f0c674',
      \ '#81a2be',
      \ '#b294bb',
      \ '#8abeb7',
      \ '#c5c8c6',
      \]

function! gin#internal#util#ansi_escape_code#colors() abort
  return s:list_colors()
endfunction

if has('nvim')
  function! s:list_colors() abort
    return map(
          \ range(16),
          \ { _, v -> get(g:, printf('terminal_color_%d', v), s:terminal_ansi_colors[v]) },
          \)
  endfunction
else
  function! s:list_colors() abort
    return get(g:, 'terminal_ansi_colors', s:terminal_ansi_colors)
  endfunction
endif
