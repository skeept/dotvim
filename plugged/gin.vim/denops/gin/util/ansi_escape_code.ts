import * as ansiEscapeCode from "https://deno.land/x/ansi_escape_code@v0.1.2/mod.ts";
import { Decoration } from "https://deno.land/x/denops_std@v3.6.0/buffer/mod.ts";
import { countVimBytes } from "./text.ts";

export function removeAnsiEscapeCode(
  text: string,
): string {
  const [t, _] = ansiEscapeCode.trimAndParse(text);
  return t;
}

export function buildDecorationsFromAnsiEscapeCode(
  content: string[],
): [string[], Decoration[]] {
  const trimmed: string[] = [];
  const decorations: Decoration[] = [];
  for (let i = 0; i < content.length; i++) {
    const [t, annons] = ansiEscapeCode.trimAndParse(content[i]);
    let previous: [number, string] | undefined;
    for (const annon of annons) {
      if (!annon.csi.sgr) {
        continue;
      }
      const highlight = highlightFromSgr(annon.csi.sgr);
      if (!highlight) {
        continue;
      }
      const byteOffset = countVimBytes(t.substring(0, annon.offset));
      if (previous) {
        decorations.push({
          line: i + 1,
          column: 1 + previous[0],
          length: byteOffset - previous[0],
          highlight: previous[1],
        });
      }
      previous = highlight === "NONE" ? undefined : [byteOffset, highlight];
    }
    if (previous) {
      decorations.push({
        line: i + 1,
        column: previous[0],
        length: countVimBytes(t) - previous[0],
        highlight: previous[1],
      });
    }
    trimmed.push(t);
  }
  return [trimmed, decorations];
}

function highlightFromSgr(sgr: ansiEscapeCode.Sgr): string | undefined {
  if (sgr.foreground === "default" || sgr.reset) {
    return "NONE";
  }
  if (sgr.foreground && typeof sgr.foreground === "number") {
    const index = (sgr.bold && sgr.foreground < 8)
      ? sgr.foreground + 8
      : sgr.foreground;
    return `GinColor${index}`;
  }
  // We don't support that SGR
  return undefined;
}
