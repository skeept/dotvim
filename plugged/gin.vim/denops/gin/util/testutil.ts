import * as path from "jsr:@std/path@^1.0.0";
import { test as testOri, TestDefinition } from "jsr:@denops/test@^3.0.0";

const runtimepath = path.resolve(
  path.fromFileUrl(new URL("../../..", import.meta.url)),
);

export function test(
  mode: TestDefinition["mode"],
  name: string,
  fn: TestDefinition["fn"],
): void;
export function test(def: TestDefinition): void;
export function test(
  modeOrDef: TestDefinition["mode"] | TestDefinition,
  name?: string,
  fn?: TestDefinition["fn"],
): void {
  if (typeof modeOrDef === "string") {
    if (!name) {
      throw new Error(`'name' attribute is required`);
    }
    if (!fn) {
      throw new Error(`'fn' attribute is required`);
    }
    test({ mode: modeOrDef, name, fn });
    return;
  }
  testOri({
    ...modeOrDef,
    prelude: [
      `set runtimepath^=${runtimepath}`,
      ...(modeOrDef.prelude ?? []),
    ],
  });
}
