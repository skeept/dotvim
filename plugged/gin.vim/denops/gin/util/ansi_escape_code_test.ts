import { assertSnapshot } from "https://deno.land/std@0.192.0/testing/snapshot.ts";
import { test } from "./testutil.ts";
import { buildDecorationsFromAnsiEscapeCode } from "./ansi_escape_code.ts";

const testText = await Deno.readTextFile(
  new URL("./testdata/ansi_escape_code_content.txt", import.meta.url),
);

test("all", "buildDecorationsFromAnsiEscapeCode", async (denops, t) => {
  await t.step({
    name: "returns proper Decoration[]",
    fn: async () => {
      const content = testText.split("\n");
      const [trimmed, decorations] = await buildDecorationsFromAnsiEscapeCode(
        denops,
        content,
      );
      await assertSnapshot(t, {
        content,
        trimmed,
        decorations,
      });
    },
  });
});
