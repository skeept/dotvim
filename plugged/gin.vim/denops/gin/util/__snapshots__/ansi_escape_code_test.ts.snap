export const snapshot = {};

snapshot[`buildDecorationsFromAnsiEscapeCode (vim) 1`] = `
{
  content: [
    "\\x1b[1mbold\\x1b[0m",
    "\\x1b[3mitalic\\x1b[0m",
    "\\x1b[4munderline\\x1b[0m",
    "\\x1b[9mstrikethrough\\x1b[0m",
    "\\x1b[31mred\\x1b[0m",
    "",
  ],
  decorations: [
    {
      column: 1,
      highlight: "GinColorBold",
      length: 4,
      line: 1,
    },
    {
      column: 1,
      highlight: "GinColorItalic",
      length: 6,
      line: 2,
    },
    {
      column: 1,
      highlight: "GinColorUnderline",
      length: 9,
      line: 3,
    },
    {
      column: 1,
      highlight: "GinColorStrike",
      length: 13,
      line: 4,
    },
    {
      column: 1,
      highlight: "GinColorF1",
      length: 3,
      line: 5,
    },
  ],
  trimmed: [
    "bold",
    "italic",
    "underline",
    "strikethrough",
    "red",
    "",
  ],
}
`;

snapshot[`buildDecorationsFromAnsiEscapeCode (nvim) 1`] = `
{
  content: [
    "\\x1b[1mbold\\x1b[0m",
    "\\x1b[3mitalic\\x1b[0m",
    "\\x1b[4munderline\\x1b[0m",
    "\\x1b[9mstrikethrough\\x1b[0m",
    "\\x1b[31mred\\x1b[0m",
    "",
  ],
  decorations: [
    {
      column: 1,
      highlight: "GinColorBold",
      length: 4,
      line: 1,
    },
    {
      column: 1,
      highlight: "GinColorItalic",
      length: 6,
      line: 2,
    },
    {
      column: 1,
      highlight: "GinColorUnderline",
      length: 9,
      line: 3,
    },
    {
      column: 1,
      highlight: "GinColorStrike",
      length: 13,
      line: 4,
    },
    {
      column: 1,
      highlight: "GinColorF1",
      length: 3,
      line: 5,
    },
  ],
  trimmed: [
    "bold",
    "italic",
    "underline",
    "strikethrough",
    "red",
    "",
  ],
}
`;
